pex/environment.py

```diff
@@ -335,29 +383,37 @@ def _evaluate_marker(
         requirement,  # type: Requirement
         extras=None,  # type: Optional[Tuple[str, ...]]
     ):
-        # type: (...) -> bool
+        # type: (...) -> Optional[bool]
         applies = self._target.requirement_applies(requirement, extras=extras)
-        if not applies:
+        if applies is False:
             TRACER.log(
                 "Skipping activation of `{}` due to environment marker de-selection".format(
                     requirement
-                )
+                ),
+                V=3,
             )
         return applies
 
     def _resolve_requirement(
         self,
         requirement,  # type: Requirement
+        resolved_dists_by_key,  # type: MutableMapping[Distribution, _RequirementKey]
+        required,  # type: Optional[bool]
         required_by=None,  # type: Optional[Distribution]
     ):
-        # type: (...) -> Iterator[Union[Distribution, _DistributionNotFound]]
+        # type: (...) -> Iterator[_DistributionNotFound]
+        requirement_key = _RequirementKey.create(requirement)
+        if requirement_key in resolved_dists_by_key:
```

Comment:
This change to populate `resolved_dists_by_key` as a side-effect and short circuit drops a large, connected resolve graph from 1.2s to 0.6s in my stress testing (and that down from 15s for a normal network-based resolve using Pip with a warm cache). In other words, just resolving the graph minimally is an important optimization that seems to justify this ugliness.